File: sdk/python/captcha-background-sdk/captcha_background_sdk/diff_engine.py
from __future__ import annotations

from typing import List, Optional, Tuple

RGB = Tuple[int, int, int]


def build_diff(
    cap_px: List[Tuple[int, ...]],
    bg_px: List[Tuple[int, ...]],
    width: int,
    height: int,
    diff_threshold: int,
) -> Tuple[List[bool], List[Optional[RGB]]]:
    total = width * height
    mask = [False] * total
    colors: List[Optional[RGB]] = [None] * total
    for i in range(total):
        cr, cg, cb, _ca = cap_px[i]
        br, bg, bb, _ba = bg_px[i]
        if abs(int(cr) - int(br)) + abs(int(cg) - int(bg)) + abs(int(cb) - int(bb)) >= diff_threshold:
            mask[i] = True
            colors[i] = (int(cr), int(cg), int(cb))
    return mask, colors

File: typescript-sdk/src/diff-engine.ts
import { RGBA, RGB } from './types';

export function buildDiff(
  captchaPixels: RGBA[],
  backgroundPixels: RGBA[],
  width: number,
  height: number,
  diffThreshold: number
): { mask: boolean[]; colors: Array<RGB | null> } {
  const total = width * height;
  const mask = new Array<boolean>(total).fill(false);
  const colors: Array<RGB | null> = new Array(total).fill(null);

  for (let i = 0; i < total; i += 1) {
    const [cr, cg, cb] = captchaPixels[i];
    const [br, bg, bb] = backgroundPixels[i];
    if (Math.abs(cr - br) + Math.abs(cg - bg) + Math.abs(cb - bb) >= diffThreshold) {
      mask[i] = true;
      colors[i] = [cr, cg, cb];
    }
  }

  return { mask, colors };
}
